# Hospital Utilization Survival Analysis

In [1]:
#import requests
#from pytrends.request import TrendReq
import pandas as pd
#from datetime import datelast_iot_date_plus_week
#from datetime import timedelta
from load_data import create_data, update_data
%load_ext autoreload # TODO: Delete this when finished

## Data

### Google Trends Data
[Pytrends API Documentation](https://pypi.org/project/pytrends/)

### Hospital and COVID Data

### HealthData.gov COVID-19 Reported Patient Impact and Hospital Capacity by Facility
[Hospital Capacity Data](https://healthdata.gov/Hospital/COVID-19-Reported-Patient-Impact-and-Hospital-Capa/anag-cw7u)

### CDC COVID-19 Case Surveillance Public Use Data with Geography
[COVID Case Surveillance Data](https://data.cdc.gov/Case-Surveillance/COVID-19-Case-Surveillance-Public-Use-Data-with-Ge/n8mc-b4w4)

[HealthData.gov API docs](https://dev.socrata.com/foundry/healthdata.gov/6xf2-c3ie)

## Create Data Sets
### If data sets have already been created, update them, or move on to exploration.

In [54]:
#TODO: Delete following 2 lines when finished
%autoreload
from load_data import update_data

# Create data
dfs_list = create_data()
# Save df's separately
reg_int_df = dfs_list[0]
iot_df = dfs_list[1]
hospitals_df = dfs_list[2]
covid_df = dfs_list[3]

In [7]:
# TODO: Create alert that an update should be performed

## Update Data
### If the data have been created already, they can be updated below.

In [125]:
#TODO: Delete following 2 lines when finished
%autoreload
from load_data import update_data

# Create data
updated_dfs_list = create_data()
# Save df's separately
reg_int_df = updated_dfs_list[0]
iot_df = updated_dfs_list[1]
hospitals_df = updated_dfs_list[2]
covid_df = updated_dfs_list[3]

In [17]:
#NOTE: Save dataframes as prior states from which present missing values can be fille?

## Explore Data
### Hospital Data

In [124]:
hospitals_df.head(1)

hospital_pk          collection_week state     ccn  \
0      050767  2021-09-03T00:00:00.000    CA  050767   

                            hospital_name          address       city    zip  \
0  KAISER FOUNDATION HOSPITAL - VACAVILLE  1 QUALITY DRIVE  VACAVILLE  95688   

  hospital_subtype  fips_code  ...  \
0       Short Term     6095.0  ...   

  previous_day_admission_adult_covid_suspected_7_day_coverage  \
0                                                  7            

   previous_day_admission_pediatric_covid_suspected_7_day_coverage  \
0                                                  7                 

   previous_week_personnel_covid_vaccinated_doses_administered_7_day  \
0                                                0.0                   

   total_personnel_covid_vaccinated_doses_none_7_day  \
0                                                0.0   

   total_personnel_covid_vaccinated_doses_one_7_day  \
0                                               0.0   

   total_personnel_covid_vaccinated_doses_all_7_day  \
0                                               0.0   

   previous_week_patients_covid_vaccinated_doses_one_7_day  \
0                                                0.0         

   previous_week_patients_covid_vaccinated_doses_all_7_day  is_corrected  \
0                                                0.0               False   

   :@computed_region_pqdx_y6mm  
0                          5.0  

[1 rows x 107 columns]

In [5]:
hosp_cols = hospitals_df.columns
hosp_cols

Index(['hospital_pk', 'collection_week', 'state', 'ccn', 'hospital_name',
       'address', 'city', 'zip', 'hospital_subtype', 'fips_code',
       ...
       'previous_day_admission_adult_covid_suspected_7_day_coverage',
       'previous_day_admission_pediatric_covid_suspected_7_day_coverage',
       'previous_week_personnel_covid_vaccinated_doses_administered_7_day',
       'total_personnel_covid_vaccinated_doses_none_7_day',
       'total_personnel_covid_vaccinated_doses_one_7_day',
       'total_personnel_covid_vaccinated_doses_all_7_day',
       'previous_week_patients_covid_vaccinated_doses_one_7_day',
       'previous_week_patients_covid_vaccinated_doses_all_7_day',
       'is_corrected', ':@computed_region_pqdx_y6mm'],
      dtype='object', length=107)

### There are 107 columns in the hospital data. We wont be using all of it. But, rather than dropping so many of the columns, let's subset it to create smaller data sets. This way, we will have the ability to use this data later.

### Let's start by focusing on the bed data.
### Bed Data

In [128]:
# Columns to keep
bed_cols = ["hospital_pk",
            "collection_week",
            "state",
            "ccn",
            "hospital_name",
            "address",
            "city",
            "zip",
            "hospital_subtype",
            "fips_code",
            "is_metro_micro",
            "total_beds_7_day_avg",
            "all_adult_hospital_beds_7_day_avg",
            "all_adult_hospital_inpatient_beds_7_day_avg",
            "inpatient_beds_used_7_day_avg",
            "all_adult_hospital_inpatient_bed_occupied_7_day_avg"]
                  
# Check data types, dimensions# Check data types, dimensions
beds_df.info()
beds_df = hospitals_df[bed_cols]
beds_df.head(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   hospital_pk                                          1000 non-null   object 
 1   collection_week                                      1000 non-null   object 
 2   state                                                1000 non-null   object 
 3   ccn                                                  997 non-null    object 
 4   hospital_name                                        1000 non-null   object 
 5   address                                              999 non-null    object 
 6   city                                                 999 non-null    object 
 7   zip                                                  1000 non-null   int64  
 8   hospital_subtype                                     1000 non-null   

hospital_pk          collection_week state     ccn          hospital_name  \
0      280023  2021-09-03T00:00:00.000    NE  280023  CHI HEALTH ST FRANCIS   

                 address          city    zip hospital_subtype  fips_code  \
0  2620 WEST FAIDLEY AVE  GRAND ISLAND  68803       Short Term    31079.0   

  is_metro_micro  total_beds_7_day_avg  all_adult_hospital_beds_7_day_avg  \
0           true                  92.0                               84.0   

   all_adult_hospital_inpatient_beds_7_day_avg  inpatient_beds_used_7_day_avg  \
0                                         80.0                           64.3   

   all_adult_hospital_inpatient_bed_occupied_7_day_avg  
0                                               63.4

In [ ]:
#TODO: I need to figure out how I'm going to handle the negtive and missing values.

In [129]:
# Descriptive statistics
beds_df.describe()

zip     fips_code  total_beds_7_day_avg  \
count   1000.000000    999.000000           1000.000000   
mean   53326.554000  30244.963964          -1826.364400   
std    27021.650725  16116.623627          44707.236758   
min      613.000000   1011.000000        -999999.000000   
25%    32380.750000  17201.000000             29.925000   
50%    54630.500000  30029.000000             79.350000   
75%    76033.000000  42131.000000            244.250000   
max    99615.000000  72141.000000           1574.000000   

       all_adult_hospital_beds_7_day_avg  \
count                        1000.000000   
mean                        -1844.552800   
std                         44706.306053   
min                       -999999.000000   
25%                            25.000000   
50%                            70.000000   
75%                           222.000000   
max                          1376.000000   

       all_adult_hospital_inpatient_beds_7_day_avg  \
count                                  1000.000000   
mean                                  -2867.365300   
std                                   54724.895766   
min                                 -999999.000000   
25%                                      23.075000   
50%                                      58.000000   
75%                                     181.750000   
max                                    1306.000000   

       inpatient_beds_used_7_day_avg  \
count                    1000.000000   
mean                   -62885.044000   
std                    243114.149438   
min                   -999999.000000   
25%                        11.900000   
50%                        43.650000   
75%                       152.700000   
max                      1321.100000   

       all_adult_hospital_inpatient_bed_occupied_7_day_avg  
count                                        1000.000000    
mean                                       -63895.464400    
std                                        244902.534405    
min                                       -999999.000000    
25%                                            11.300000    
50%                                            40.500000    
75%                                           143.950000    
max                                          1189.900000

In [7]:
# Save the number of observations
num_obs = len(hospitals_df)

# Drop rows if na's are in bed data columns
hospitals_df.drop(hospitals_df.iloc[11:15,:].isna(), index=True, axis=1)
hospitals_df.drop(hospitals_df.iloc[11:15,:] < 0, index=True, axis=1)

# Print the number of remaining observations
print("The remaining number of observations after dropping rows with na's is: ", num_obs - len(hospitals_df))

ValueError: Cannot specify both 'labels' and 'index'/'columns'

In [ ]:
# Get descriptive statistics
hospitals_df.describe()

zip     fips_code  total_beds_7_day_avg  \
count   1000.000000   1000.000000           1000.000000   
mean   53039.459000  29511.847000           -829.921600   
std    26086.456729  15765.141093          31629.149148   
min      613.000000   1007.000000        -999999.000000   
25%    32481.750000  18035.500000             27.100000   
50%    53814.000000  28101.000000             65.000000   
75%    74874.250000  42051.000000            218.275000   
max    99901.000000  72127.000000           3168.000000   

       all_adult_hospital_beds_7_day_avg  \
count                        1000.000000   
mean                         -849.058700   
std                         31628.295842   
min                       -999999.000000   
25%                            25.000000   
50%                            60.150000   
75%                           190.000000   
max                          2573.000000   

       all_adult_hospital_inpatient_beds_7_day_avg  \
count                                  1000.000000   
mean                                   -871.126100   
std                                   31627.419056   
min                                 -999999.000000   
25%                                      22.100000   
50%                                      50.000000   
75%                                     160.500000   
max                                    2329.000000   

       inpatient_beds_used_7_day_avg  \
count                    1000.000000   
mean                   -66892.168700   
std                    250175.810833   
min                   -999999.000000   
25%                        10.700000   
50%                        35.850000   
75%                       132.025000   
max                      2384.600000   

       all_adult_hospital_inpatient_bed_occupied_7_day_avg  
count                                        1000.000000    
mean                                       -68901.990900    
std                                        253607.432743    
min                                       -999999.000000    
25%                                             9.700000    
50%                                            33.200000    
75%                                           125.150000    
max                                          1974.700000

### Above, we can see that there are some values worth investigating. Each of the columns that contain bed counts have minimum values of -999,999.0. The means of these columns are negative, as weell. It remains to be seen if that is because of the large negative minimum value/s.

In [8]:
hospitals_df.query('total_beds_7_day_avg < 0')

Empty DataFrame
Columns: [hospital_pk, collection_week, state, ccn, hospital_name, address, city, zip, hospital_subtype, fips_code, is_metro_micro, total_beds_7_day_avg, all_adult_hospital_beds_7_day_avg, all_adult_hospital_inpatient_beds_7_day_avg, inpatient_beds_used_7_day_avg, all_adult_hospital_inpatient_bed_occupied_7_day_avg, total_adult_patients_hospitalized_confirmed_and_suspected_covid_7_day_avg, total_adult_patients_hospitalized_confirmed_covid_7_day_avg, total_pediatric_patients_hospitalized_confirmed_and_suspected_covid_7_day_avg, total_pediatric_patients_hospitalized_confirmed_covid_7_day_avg, inpatient_beds_7_day_avg, total_icu_beds_7_day_avg, total_staffed_adult_icu_beds_7_day_avg, icu_beds_used_7_day_avg, staffed_adult_icu_bed_occupancy_7_day_avg, staffed_icu_adult_patients_confirmed_and_suspected_covid_7_day_avg, staffed_icu_adult_patients_confirmed_covid_7_day_avg, total_patients_hospitalized_confirmed_influenza_7_day_avg, icu_patients_confirmed_influenza_7_day_avg, total_patients_hospitalized_confirmed_influenza_and_covid_7_day_avg, total_beds_7_day_sum, all_adult_hospital_beds_7_day_sum, all_adult_hospital_inpatient_beds_7_day_sum, inpatient_beds_used_7_day_sum, all_adult_hospital_inpatient_bed_occupied_7_day_sum, total_adult_patients_hospitalized_confirmed_and_suspected_covid_7_day_sum, total_adult_patients_hospitalized_confirmed_covid_7_day_sum, total_pediatric_patients_hospitalized_confirmed_and_suspected_covid_7_day_sum, total_pediatric_patients_hospitalized_confirmed_covid_7_day_sum, inpatient_beds_7_day_sum, total_icu_beds_7_day_sum, total_staffed_adult_icu_beds_7_day_sum, icu_beds_used_7_day_sum, staffed_adult_icu_bed_occupancy_7_day_sum, staffed_icu_adult_patients_confirmed_and_suspected_covid_7_day_sum, staffed_icu_adult_patients_confirmed_covid_7_day_sum, total_patients_hospitalized_confirmed_influenza_7_day_sum, icu_patients_confirmed_influenza_7_day_sum, total_patients_hospitalized_confirmed_influenza_and_covid_7_day_sum, total_beds_7_day_coverage, all_adult_hospital_beds_7_day_coverage, all_adult_hospital_inpatient_beds_7_day_coverage, inpatient_beds_used_7_day_coverage, all_adult_hospital_inpatient_bed_occupied_7_day_coverage, total_adult_patients_hospitalized_confirmed_and_suspected_covid_7_day_coverage, total_adult_patients_hospitalized_confirmed_covid_7_day_coverage, total_pediatric_patients_hospitalized_confirmed_and_suspected_covid_7_day_coverage, total_pediatric_patients_hospitalized_confirmed_covid_7_day_coverage, inpatient_beds_7_day_coverage, total_icu_beds_7_day_coverage, total_staffed_adult_icu_beds_7_day_coverage, icu_beds_used_7_day_coverage, staffed_adult_icu_bed_occupancy_7_day_coverage, staffed_icu_adult_patients_confirmed_and_suspected_covid_7_day_coverage, staffed_icu_adult_patients_confirmed_covid_7_day_coverage, total_patients_hospitalized_confirmed_influenza_7_day_coverage, icu_patients_confirmed_influenza_7_day_coverage, total_patients_hospitalized_confirmed_influenza_and_covid_7_day_coverage, previous_day_admission_adult_covid_confirmed_7_day_sum, previous_day_admission_adult_covid_confirmed_18_19_7_day_sum, previous_day_admission_adult_covid_confirmed_20_29_7_day_sum, previous_day_admission_adult_covid_confirmed_30_39_7_day_sum, previous_day_admission_adult_covid_confirmed_40_49_7_day_sum, previous_day_admission_adult_covid_confirmed_50, previous_day_admission_adult_covid_confirmed_60, previous_day_admission_adult_covid_confirmed_70, previous_day_admission_adult_covid_confirmed_80, previous_day_admission_adult_covid_confirmed_unknown_7_day_sum, previous_day_admission_pediatric_covid_confirmed_7_day_sum, previous_day_covid_ed_visits_7_day_sum, previous_day_admission_adult_covid_suspected_7_day_sum, previous_day_admission_adult_covid_suspected_18, previous_day_admission_adult_covid_suspected_20, previous_day_admission_adult_covid_suspected_30, previous_day_admission_adult_covid_suspected_40, previous_day_admission_adult_covid_suspected_50, previous_day_admission_adult_cov

### Above are the two rows that contain negative values in the total_beds_7_day_avg column. We can see that there are other negtive values in those rows, too. Let's drop them because they do not provide quality information.

In [9]:
hospitals_df.loc[:, 'total_beds_7_day_avg':'all_adult_hospital_inpatient_bed_occupied_7_day_avg']
#hospitals_df.get_iloc('total_beds_7_day_avg:all_adult_hospital_inpatient_bed_occupied_7_day_avg')



#[hospitals_df.columns.get_loc(c) for c in cols if c in hosptials_df]

total_beds_7_day_avg  all_adult_hospital_beds_7_day_avg  \
0                   398.1                              398.1   
1                    51.0                               51.0   
2                     6.9                                6.9   
3                    88.0                               65.0   
4                   146.0                              146.0   
..                    ...                                ...   
995                  57.4                               57.4   
996                 197.0                              193.0   
997                  57.0                               57.0   
998                 401.0                              401.0   
999                  19.0                               19.0   

     all_adult_hospital_inpatient_beds_7_day_avg  \
0                                          227.3   
1                                           45.0   
2                                            6.9   
3                                           65.0   
4                                          117.0   
..                                           ...   
995                                         57.4   
996                                        142.0   
997                                         57.0   
998                                        248.7   
999                                    -999999.0   

     inpatient_beds_used_7_day_avg  \
0                            205.4   
1                             19.4   
2                              5.1   
3                             35.3   
4                             91.4   
..                             ...   
995                           38.3   
996                           91.1   
997                           11.0   
998                          233.4   
999                      -999999.0   

     all_adult_hospital_inpatient_bed_occupied_7_day_avg  
0                                                177.7    
1                                                 19.4    
2                                                  5.1    
3                                                 34.0    
4                                                 91.4    
..                                                 ...    
995                                               38.3    
996                                               88.9    
997                                               11.0    
998                                              233.4    
999                                          -999999.0    

[1000 rows x 5 columns]

In [10]:
# Save the row index, drop those rows, check number of remaining observations
rows = hospitals_df.query('total_beds_7_day_avg < 0').index
hospitals_df.drop(rows, inplace=True)
print(len(hospitals_df))

1000


In [11]:
# Get descriptive statistics
hospitals_df.describe()

zip    fips_code  total_beds_7_day_avg  \
count   1000.000000   1000.00000           1000.000000   
mean   54693.932000  28845.68000            157.065400   
std    26775.387353  15600.60776            212.308806   
min      682.000000   1003.00000              0.000000   
25%    35054.750000  17120.50000             26.000000   
50%    56480.500000  28012.00000             63.200000   
75%    76497.750000  42003.00000            215.250000   
max    99801.000000  72125.00000           1977.000000   

       all_adult_hospital_beds_7_day_avg  \
count                        1000.000000   
mean                          142.341000   
std                           191.882975   
min                             0.000000   
25%                            25.000000   
50%                            58.500000   
75%                           190.850000   
max                          1878.000000   

       all_adult_hospital_inpatient_beds_7_day_avg  \
count                                  1000.000000   
mean                                  -4879.855200   
std                                   70577.610542   
min                                 -999999.000000   
25%                                      22.000000   
50%                                      49.850000   
75%                                     160.775000   
max                                    1408.000000   

       inpatient_beds_used_7_day_avg  \
count                    1000.000000   
mean                   -75899.198700   
std                    265159.389008   
min                   -999999.000000   
25%                        10.000000   
50%                        35.650000   
75%                       130.750000   
max                      1179.900000   

       all_adult_hospital_inpatient_bed_occupied_7_day_avg  \
count                                        1000.000000     
mean                                       -76906.981400     
std                                        266751.467883     
min                                       -999999.000000     
25%                                             9.100000     
50%                                            31.650000     
75%                                           124.625000     
max                                          1079.100000     

       total_adult_patients_hospitalized_confirmed_and_suspected_covid_7_day_avg  \
count                                        1000.000000                           
mean                                      -265982.447100                           
std                                        442095.301025                           
min                                       -999999.000000                           
25%                                       -999999.000000                           
50%                                             5.900000                           
75%                                            22.925000                           
max                                           210.700000                           

       total_adult_patients_hospitalized_confirmed_covid_7_day_avg  \
count                                        1000.000000             
mean                                      -256983.408000             
std                                        437207.233672             
min                                       -999999.000000             
25%                                       -999999.000000             
50%                                             5.500000             
75%                                            21.100000             
max                                           208.900000             

       total_pediatric_patients_hospitalized_confirmed_and_suspected_covid_7_day_avg  \
count                                        1000.000000                               
mean                                      -107999.667000                               
std                                  

### COVID-19 Data

In [13]:
# Check data types, dimensions
covid_df.info()# Check data types, dimensions
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   case_month                1000 non-null   object 
 1   res_state                 1000 non-null   object 
 2   state_fips_code           1000 non-null   int64  
 3   res_county                1000 non-null   object 
 4   county_fips_code          1000 non-null   object 
 5   age_group                 1000 non-null   object 
 6   sex                       1000 non-null   object 
 7   race                      1000 non-null   object 
 8   ethnicity                 1000 non-null   object 
 9   case_positive_specimen    348 non-null    float64
 10  process                   1000 non-null   object 
 11  exposure_yn               1000 non-null   object 
 12  current_status            1000 non-null   object 
 13  symptom_status            1000 non-null   object 
 14  hosp_yn  

In [11]:
# Check for missing values
covid_df.isna().sum().sum()

2108

In [15]:
# Get descriptive statistics
covid_df.describe()

state_fips_code  case_positive_specimen  case_onset_interval
count      1000.000000              348.000000           484.000000
mean         29.584000                0.255747             0.099174
std          15.242483                2.589181             1.425325
min           1.000000              -27.000000             0.000000
25%          18.000000                0.000000             0.000000
50%          29.000000                0.000000             0.000000
75%          42.000000                0.000000             0.000000
max          55.000000               31.000000            24.000000